File: app/src/main/java/com/example/quoter/utils/ImageUtils.kt
package com.example.quoter.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Rect
import java.io.BufferedInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object ImageUtils {

    /**
     * Crops a bitmap to a 9:20 aspect ratio, keeping the center.
     * Assumes the input bitmap is 1024x1792 (or similar aspect ratio from DALL-E 3).
     */
    fun cropTo9x16(source: Bitmap): Bitmap? { // Keep name for compatibility, but ratio is 9:20
        val sourceWidth = source.width
        val sourceHeight = source.height

        // Target aspect ratio (9:20)
        val targetAspectRatio = 9.0 / 20.0 // New ratio

        // Calculate the target width based on source height to maintain 9:20
        var targetWidth = (sourceHeight * targetAspectRatio).toInt()
        var targetHeight = sourceHeight

        // If the calculated width is wider than the source, calculate target height based on source width
        if (targetWidth > sourceWidth) {
            targetWidth = sourceWidth
            targetHeight = (sourceWidth / targetAspectRatio).toInt()
            // Ensure targetHeight doesn't exceed sourceHeight
            if (targetHeight > sourceHeight) targetHeight = sourceHeight
            println("Cropping: Using full width ($sourceWidth), calculated height ($targetHeight) for 9:20 ratio.")
        } else {
             println("Cropping: Using full height ($sourceHeight), calculated width ($targetWidth) for 9:20 ratio.")
        }

        // Calculate cropping bounds (centered)
        val left = (sourceWidth - targetWidth) / 2
        val top = (sourceHeight - targetHeight) / 2
        val right = left + targetWidth
        val bottom = top + targetHeight

        // Ensure bounds are within the source bitmap dimensions
        if (left < 0 || top < 0 || right > sourceWidth || bottom > sourceHeight || targetWidth <= 0 || targetHeight <= 0) {
            println("Error: Calculated crop bounds are invalid.")
            return null // Or return the original bitmap if cropping fails
        }

        return try {
            Bitmap.createBitmap(source, left, top, targetWidth, targetHeight)
        } catch (e: Exception) {
            println("Error cropping bitmap: ${e.message}")
            null
        }
    }

    /**
     * Saves a bitmap to the app's internal files directory.
     * Returns the absolute path of the saved file, or null on failure.
     */
    fun saveBitmapToInternalStorage(context: Context, bitmap: Bitmap, filenamePrefix: String = "wallpaper_"): String? {
        val directory = context.filesDir // App's internal files directory
        // Create a unique filename
        val filename = "${filenamePrefix}${UUID.randomUUID()}.png"
        val file = File(directory, filename)

        var fileOutputStream: FileOutputStream? = null
        return try {
            fileOutputStream = FileOutputStream(file)
            // Compress the bitmap to the output stream (PNG format)
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, fileOutputStream)
            file.absolutePath // Return the full path to the saved file
        } catch (e: IOException) {
            println("Error saving bitmap: ${e.message}")
            e.printStackTrace()
            null
        } finally {
            try {
                fileOutputStream?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }
}

File: swift_files/ImageUtils.swift
swift
import UIKit

class ImageUtils {

    /**
     Crops a UIImage to a 9:20 aspect ratio, keeping the center.
     Assumes the input image is 1024x1792 (or similar aspect ratio from DALL-E 3).
     */
    class func cropTo9x20(source: UIImage) -> UIImage? {
        let sourceWidth = source.size.width
        let sourceHeight = source.size.height

        // Target aspect ratio (9:20)
        let targetAspectRatio = 9.0 / 20.0

        // Calculate the target width based on source height to maintain 9:20
        var targetWidth = Int(sourceHeight * targetAspectRatio)
        var targetHeight = Int(sourceHeight)

        // If the calculated width is wider than the source, calculate target height based on source width
        if CGFloat(targetWidth) > sourceWidth {
            targetWidth = Int(sourceWidth)
            targetHeight = Int(sourceWidth / targetAspectRatio)
            // Ensure targetHeight doesn't exceed sourceHeight
            if CGFloat(targetHeight) > sourceHeight { targetHeight = Int(sourceHeight) }
            print("Cropping: Using full width (\(sourceWidth)), calculated height (\(targetHeight)) for 9:20 ratio.")
        } else {
            print("Cropping: Using full height (\(sourceHeight)), calculated width (\(targetWidth)) for 9:20 ratio.")
        }

        // Calculate cropping bounds (centered)
        let left = (Int(sourceWidth) - targetWidth) / 2
        let top = (Int(sourceHeight) - targetHeight) / 2
        let right = left + targetWidth
        let bottom = top + targetHeight

        // Ensure bounds are within the source image dimensions
        if left < 0 || top < 0 || right > Int(sourceWidth) || bottom > Int(sourceHeight) || targetWidth <= 0 || targetHeight <= 0 {
            print("Error: Calculated crop bounds are invalid.")
            return nil // Or return the original image if cropping fails
        }

        // Create the cropping rectangle
        let cropRect = CGRect(x: CGFloat(left), y: CGFloat(top), width: CGFloat(targetWidth), height: CGFloat(targetHeight))

        // Perform the cropping
        if let croppedCGImage = source.cgImage?.cropping(to: cropRect) {
            return UIImage(cgImage: croppedCGImage, scale: source.scale, orientation: source.imageOrientation)
        } else {
            print("Error cropping image.")
            return nil
        }
    }

    /**
     Saves a UIImage to the app's documents directory.
     Returns the URL of the saved file, or nil on failure.
     */
    class func saveImageToDocumentsDirectory(image: UIImage, filenamePrefix: String = "wallpaper_") -> URL? {
        guard let data = image.pngData() else {
            print("Error: Could not convert image to PNG data.")
            return nil
        }

        // Get the documents directory URL
        guard let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            print("Error: Could not access documents directory.")
            return nil
        }

        // Create a unique filename
        let filename = "\(filenamePrefix)\(UUID().uuidString).png"
        let fileURL = documentsDirectory.appendingPathComponent(filename)

        do {
            try data.write(to: fileURL)
            return fileURL
        } catch {
            print("Error saving image: \(error.localizedDescription)")
            return nil
        }
    }
}